manifest.go

```diff
@@ -324,6 +325,71 @@ func ValidateProjectRoots(c *Ctx, m *Manifest, sm gps.SourceManager) error {
 	return valErr
 }
 
+// ValidatePackageRules ensure that there are no ineffectual package declarations
+// in the constraints, overrides, and ignored rules in the manifest.
+func ValidatePackageRules(c *Ctx, proj *Project, sm gps.SourceManager) (err error) {
```

Comment:
What about separating responsibilities here by dropping the `c *Ctx` argument and instead returning an extended `errInefficientRules` error that knows how to print itself the same way?

Reply:
I like the idea, of dropping `Ctx`, but to clarify do you mean appending more lines to the `errInefficentRules` before returning it?

Reply:
I was thinking expand `var errInefficentRules error` from a single static error instance to something like `type errInefficentRules struct{constraints, ignores []gps.ProjectRoot}` with an `Error() string` method (which e.g. builds a string via `bytes.Buffer` rather than logging).  So, essentially the same result as adding lines to `errInefficentRules` *before* returning, but underneath it would actually be deferred until the `error` is printed.

Reply:
^ This would also set us up to potentially export `errInefficentRules` and its fields later on, giving callers even more flexibility when interpreting and handling the error.

Reply:
Yeah that sounds pretty good to me, will do